controller/utils.go

```diff
@@ -122,104 +122,119 @@ func replacePodsAndWait(
 	goroutinePoolSize int,
 ) (timedout, canceled bool, err error) {
 	logger.Debug("starting to replace pods with new ones")
-	stop := make(chan struct{}, 1)
-	defer close(stop) // Dont leak goroutines
-	finishedReplace := make(chan struct{})
 
-	timedoutChan := make(chan bool)
-	canceledChan := make(chan bool)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	finishedReplace := make(chan struct{})
+	canceledChan := make(chan struct{})
 	errChan := make(chan error)
 
 	pods := make(chan *models.Pod, len(podsChunk))
 	for _, pod := range podsChunk {
 		pods <- pod
 	}
 
+	inRollingUpdate := operationManager != nil
+
+	var wg sync.WaitGroup
 	logger.Infof("starting %d in-memory workers to replace %d pods", goroutinePoolSize, len(podsChunk))
 	for i := 0; i < goroutinePoolSize; i++ {
-		go replacePodWorker(
-			logger,
-			roomManager,
-			mr,
-			clientset,
-			db,
-			redisClient,
-			willTimeoutAt,
-			configYAML,
-			scheduler,
-			operationManager,
-			clock,
-			pods,
-			stop, finishedReplace,
-			timedoutChan, canceledChan, errChan,
-		)
+		wg.Add(1)
+		go func () {
+			defer wg.Done()
+			replacePodWorker(
+				ctx,
+				logger,
+				roomManager,
+				mr,
+				clientset,
+				db,
+				redisClient,
+				configYAML,
+				scheduler,
+				pods,
+				inRollingUpdate,
+				finishedReplace,
+				errChan,
+			)
+		}()
+	}
+	
+	if inRollingUpdate {
+		go func() {
+			for {
+				canceled, err := operationManager.WasCanceled()
+				if err != nil {
+					continue
+				}
+				if canceled {
+					close(canceledChan)
+					return
+				}
+			}
+		}()
 	}
 
+	duration := willTimeoutAt.Sub(clock.Now())
+	timeout := time.NewTimer(duration)
+
 	select {
 	case err = <-errChan:
-		return false, false, err
-
-	case canceled = <-canceledChan:
-		return false, canceled, nil
-
-	case timedout = <-timedoutChan:
-		return timedout, false, nil
-
+		logger.Error("operation terminated with error")
+	case <-canceledChan:
+		logger.Warn("operation was canceled")
+		canceled = true
+	case <-timeout.C:
+		logger.Warn("operation timedout")
+		timedout = true
 	case <-finishedReplace:
 		logger.Debug("all pods were successfully replaced")
-		break
 	}
 
+	cancel()
+	wg.Wait()
+
 	return timedout, canceled, err
 }
 
 func replacePodWorker(
+	ctx context.Context,
 	logger logrus.FieldLogger,
 	roomManager models.RoomManager,
 	mr *models.MixedMetricsReporter,
 	clientset kubernetes.Interface,
 	db pginterfaces.DB,
 	redisClient redisinterfaces.RedisClient,
-	willTimeoutAt time.Time,
 	configYAML *models.ConfigYAML,
 	scheduler *models.Scheduler,
-	operationManager *models.OperationManager,
-	clock clockinterfaces.Clock,
 	pods <-chan *models.Pod,
```

Comment:
On the function above pods is an array, however in this function is a channel, is this valid in go?

Reply:
oh ok, didn't see that.